include/ginkgo/core/solver/batch_solver_base.hpp

```diff
@@ -0,0 +1,297 @@
+{
+    return {params.preconditioner, params.generated_preconditioner,
+            params.default_residual_tol, params.default_max_iterations};
+}
+
+
+}  // namespace detail
+
+
+struct BatchInfo {
+    std::unique_ptr<log::BatchLogDataBase> logdata;
+};
+
+
+/**
+ * @tparam PolymorphicBase  The base class; must be a subclass of BatchLinOp.
+ */
+template <typename ConcreteSolver,
+          typename ValueType = typename ConcreteSolver::value_type,
+          typename PolymorphicBase = BatchLinOp>
+class EnableBatchSolver
+    : public BatchSolver,
+      public EnableBatchLinOp<ConcreteSolver, PolymorphicBase> {
+public:
+    const ConcreteSolver* apply(ptr_param<const MultiVector<ValueType>> b,
+                                ptr_param<MultiVector<ValueType>> x) const
+    {
+        this->validate_application_parameters(b.get(), x.get());
+        auto exec = this->get_executor();
+        this->apply_impl(make_temporary_clone(exec, b).get(),
+                         make_temporary_clone(exec, x).get());
+        return self();
+    }
+
+    const ConcreteSolver* apply_impl(
+        ptr_param<const MultiVector<ValueType>> alpha,
+        ptr_param<const MultiVector<ValueType>> b,
+        ptr_param<const MultiVector<ValueType>> beta,
+        ptr_param<MultiVector<ValueType>> x) const
+    {
+        this->validate_application_parameters(alpha.get(), b.get(), beta.get(),
+                                              x.get());
+        auto exec = this->get_executor();
+        this->apply_impl(make_temporary_clone(exec, alpha).get(),
+                         make_temporary_clone(exec, b).get(),
+                         make_temporary_clone(exec, beta).get(),
+                         make_temporary_clone(exec, x).get());
+        return self();
+    }
+
+    ConcreteSolver* apply(ptr_param<const MultiVector<ValueType>> b,
+                          ptr_param<MultiVector<ValueType>> x)
+    {
+        static_cast<const ConcreteSolver*>(this)->apply(b, x);
+        return self();
+    }
+
+    ConcreteSolver* apply_impl(ptr_param<const MultiVector<ValueType>> alpha,
+                               ptr_param<const MultiVector<ValueType>> b,
+                               ptr_param<const MultiVector<ValueType>> beta,
+                               ptr_param<MultiVector<ValueType>> x)
+    {
+        static_cast<const ConcreteSolver*>(this)->apply(alpha, b, beta, x);
+        return self();
+    }
+
+protected:
+    GKO_ENABLE_SELF(ConcreteSolver);
+
+    explicit EnableBatchSolver(std::shared_ptr<const Executor> exec)
+        : EnableBatchLinOp<ConcreteSolver, PolymorphicBase>(std::move(exec))
+    {}
+
+    explicit EnableBatchSolver(std::shared_ptr<const Executor> exec,
+                               std::shared_ptr<const BatchLinOp> system_matrix,
+                               detail::common_batch_params common_params)
+        : BatchSolver(system_matrix, nullptr, common_params.residual_tolerance,
```

Comment:
Similar to solver_base class, there is no cpp file for this, and I dont think it is necessary here.

Reply:
We decide if there is a .cpp or not, so we can change that. I think a good rule of thumb is to put any implementation longer than a single line into the .cpp. Otherwise, they can clutter the interface files too much.

Reply:
I think it makes sense to have mixins like these implemented in the header files directly without duplicating it again in cpp files. For concrete classes, I agree that we should probably move everything into cpp files.

Reply:
Sorry, I didn't realize we were in a templated class here.